src/System.Security.Cryptography.Cng/src/System/Security/Cryptography/CngKey.Import.cs

```diff
@@ -35,9 +47,61 @@ public static CngKey Import(byte[] keyBlob, CngKeyBlobFormat format, CngProvider
 
             SafeNCryptProviderHandle providerHandle = provider.OpenStorageProvider();
             SafeNCryptKeyHandle keyHandle;
-            ErrorCode errorCode = Interop.NCrypt.NCryptImportKey(providerHandle, IntPtr.Zero, format.Format, IntPtr.Zero, out keyHandle, keyBlob, keyBlob.Length, 0);
-            if (errorCode != ErrorCode.ERROR_SUCCESS)
-                throw errorCode.ToCryptographicException();
+            ErrorCode errorCode;
+            
+            if (curve == null)
+            {
+                errorCode = Interop.NCrypt.NCryptImportKey(providerHandle, IntPtr.Zero, format.Format, IntPtr.Zero, out keyHandle, keyBlob, keyBlob.Length, 0);
+                if (errorCode != ErrorCode.ERROR_SUCCESS)
+                {
+                    throw errorCode.ToCryptographicException();
+                }
+            }
+            else
+            {
+                // Call with Oid.FriendlyName because .Value will result in an invalid parameter error
+                Debug.Assert(curve.Value.IsNamed);
+                string curveName = curve.Value.Oid.FriendlyName; 
+                using (SafeUnicodeStringHandle safeCurveName = new SafeUnicodeStringHandle(curveName))
+                {
+                    var desc = new Interop.BCrypt.BCryptBufferDesc();
+                    var buff = new Interop.BCrypt.BCryptBuffer();
+
+                    IntPtr descPtr = IntPtr.Zero;
+                    IntPtr buffPtr = IntPtr.Zero;
+                    try
+                    {
+                        descPtr = Marshal.AllocHGlobal(Marshal.SizeOf(desc));
+                        buffPtr = Marshal.AllocHGlobal(Marshal.SizeOf(buff));
+                        buff.cbBuffer = (curveName.Length + 1) * 2; // Add 1 for null terminator
+                        buff.BufferType = Interop.BCrypt.NCryptBufferDescriptors.NCRYPTBUFFER_ECC_CURVE_NAME;
+                        buff.pvBuffer = safeCurveName.DangerousGetHandle();
+                        Marshal.StructureToPtr(buff, buffPtr, false);
+
+                        desc.cBuffers = 1;
+                        desc.pBuffers = buffPtr;
+                        desc.ulVersion = Interop.BCrypt.BCRYPTBUFFER_VERSION;
+                        Marshal.StructureToPtr(desc, descPtr, false);
+
+                        errorCode = Interop.NCrypt.NCryptImportKey(providerHandle, IntPtr.Zero, format.Format, descPtr, out keyHandle, keyBlob, keyBlob.Length, 0);
+                    }
+                    finally
+                    {
+                        Marshal.FreeHGlobal(descPtr);
```

Comment:
It can be null - FreeHGlobal will ignore it
https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.marshal.freehglobal(v=vs.110).aspx